src/declarativebookmarkmodel.cpp

```diff
@@ -36,33 +37,47 @@ QHash<int, QByteArray> DeclarativeBookmarkModel::roleNames() const
     return roles;
 }
 
-void DeclarativeBookmarkModel::addBookmark(const QString& url, const QString& title, const QString& favicon, bool touchIcon)
+void DeclarativeBookmarkModel::add(const QString& url, const QString& title, const QString& favicon, bool touchIcon)
 {
     beginInsertRows(QModelIndex(), rowCount(), rowCount());
     bookmarkIndexes.insert(url, bookmarks.count());
     bookmarks.append(new Bookmark(title, url, favicon, touchIcon));
     endInsertRows();
-
     emit countChanged();
+    // Getter will check if active page is still bookmarked.
+    emit activeUrlBookmarkedChanged();
+
     save();
 }
 
-void DeclarativeBookmarkModel::removeBookmark(const QString& url)
+void DeclarativeBookmarkModel::remove(const QString& url)
 {
     if (!contains(url)) {
         return;
     }
 
-    if (bookmarkIndexes.keys().contains(url)) {
-        int index = bookmarkIndexes.value(url);
+    int index = bookmarkIndexes.value(url, -1);
+    remove(index);
+}
+
+void DeclarativeBookmarkModel::remove(int index)
+{
+    if (index >= 0 && index < bookmarks.count()) {
         beginRemoveRows(QModelIndex(), index, index);
         Bookmark* bookmark = bookmarks.takeAt(index);
         delete bookmark;
 
         // Remove index mapping and update remaining indices
-        bookmarkIndexes.remove(url);
         QMap<QString, int>::iterator i = bookmarkIndexes.begin();
         while (i != bookmarkIndexes.end()) {
+            if (i.value() == index) {
+                i = bookmarkIndexes.erase(i);
+                // Erased item was the last one.
```

Comment:
Rebased and added a fix so that we don't crash when removing bookmarks one by one and last is removed. Also a unit test exists for the case.